src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ConstructorsDeclarationGroupingCheck.java

```diff
@@ -0,0 +1,109 @@
+///////////////////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code and other text files for adherence to a set of rules.
+// Copyright (C) 2001-2024 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+///////////////////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.coding;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import com.puppycrawl.tools.checkstyle.FileStatefulCheck;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+
+/**
+ * <p>
+ * Checks that all constructors are grouped together.
+ * If there is anything between overloaded constructors ( expect comments )
+ * then this check will give an error.
+ * </p>
+ * <p>
+ * Parent is {@code com.puppycrawl.tools.checkstyle.TreeWalker}
+ * </p>
+ * <p>
+ * Violation Message Keys:
+ * </p>
+ * <ul>
+ * <li>
+ * {@code constructors.declaration.grouping}
+ * </li>
+ * </ul>
+ *
+ * @since 10.16.0
+ */
+
+@FileStatefulCheck
+public class ConstructorsDeclarationGroupingCheck extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_KEY = "constructors.declaration.grouping";
+
+    /**
+     * Specifies different Object Blocks scope.
+     */
+    private final Map<DetailAST, Integer> allObjBlocks = new HashMap<>();
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {
+            TokenTypes.CTOR_DEF,
+            TokenTypes.COMPACT_CTOR_DEF,
+        };
+    }
+
+    @Override
+    public void beginTree(DetailAST rootAst) {
+        allObjBlocks.clear();
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        final DetailAST currObjBlock = ast.getParent();
+        final Integer previousCtorLineNo = allObjBlocks.get(currObjBlock);
```

Comment:
Why we can not simply check previous sibling on each ctor? 
We might need some collection to keep track if ctor is first or not. This collection will just track that ctor is detected, so all other ctors should be in siblings.

Reply:
This will work in most of the situations. We can keep track if the current CTOR is first or not and check for its previous sibling, but if there is another block between the constructors then this approach will not work. For example:

https://github.com/checkstyle/checkstyle/blob/5c4be0c514d0a63aa89d6e6ea711c4d0bf294632/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/constructorsdeclarationgrouping/InputConstructorsDeclarationGrouping.java#L29-L53

Here, the `Testing(double d) {}` occurs after an inner class and based on our check's implementation we will loose track of the previous CTOR and its position if we use a normal variable like a boolean or something to keep track of the previous constructor as it will get reset. That's why I used HashMap to keep track of all blocks and see if there is a constructor occurred in current block or not. We're getting the previous CTOR's line number of the current block and checking if it is null or not, null means there is no previous CTOR in the current block. If there is no CTOR occurred, then put the current block in the HashMap and store the constructor's line number. But if there is a constructor already occurred in the current block, then we check for the previous sibling, if the previous sibling is not a constructor then we will log the violation and if the previous sibling is a constructor then we will update the previous constructor's line number to current constructor's line number.

And we have provided the last occurred constructor's line number in the violation so user can know where the previous constructor is located. @romani

Reply:
Ok

Reply:
Looks like [this](https://github.com/checkstyle/checkstyle/pull/14749#discussion_r1583033202) approach might also mitigate the issue here.

Reply:
Yeah by using that approach we will be operating the check statelessly but that approach will be more time consuming, am I right? I've given my opinion [here](https://github.com/checkstyle/checkstyle/pull/14749#discussion_r1583106507)